### PR TITLE
Checkpoint

### PR DESCRIPTION
**Pull request by ekr**

### CODE REVIEW

File: draft-pda-protocol.md [hunk 1]
```diff
@@ -146,13 +146,164 @@ document.
...
+                    |            |
++--------+          |   Helper   |
+|        |          |            |
+| Client +----+     +-----^------+
+|        |    |           |
++--------+    |           |
+              |           |
++--------+    |     +-----v------+         +-----------+
+|        |    +----->            |         |           |
+| Client +---------->   Leader   +---------> Collector |
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
| Client +---------->   Leader   <---------> Collector |
```

File: draft-pda-protocol.md [hunk 2]
```diff
@@ -146,13 +146,164 @@ document.
...
+
+
+The main participants in the protocol are as follows:
+
+Collector:
+: The entity which wants to take the measurement and ultimately receives
+  the results. Any given measurement will have a single collector.
+
+Client(s):
+: The endpoints which directly take the measurement(s) and report it to the
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
: The endpoints which directly take the measurement(s) and report them to the
```

File: draft-pda-protocol.md [hunk 3]
```diff
@@ -146,13 +146,164 @@ document.
...
+  PPM system. In order to provide reasonable levels of privacy, there
+  must be a large number of clients.
```
Comment (chris-wood) [inline draft-pda-protocol.md:222]:
```suggestion
  must be a large number of clients.
```

File: draft-pda-protocol.md [hunk 4]
```diff
@@ -146,13 +146,164 @@ document.
...
+
+Aggregator:
+: An endpoint which receives report shares. Each aggregator works with the
+  other aggregators to compute the final aggregate. This protocol defines
+  two types of aggregators: a Leader and a set of Helpers.
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
  two types of aggregators: Leaders and Helpers. For each measurement,
  there is a single leader and helper.
```

File: draft-pda-protocol.md [hunk 5]
```diff
@@ -146,13 +146,164 @@ document.
...
+
+Leader:
+: The leader is responsible for coordinating the protocol. It receives
+  the encrypted shares, distributes them to the helpers and orchestrates
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
  the encrypted shares, distributes them to the helpers, and orchestrates
```

File: draft-pda-protocol.md [hunk 6]
```diff
@@ -146,13 +146,164 @@ document.
...
+  the process of computing the final measurement.
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
  the process of computing the final measurement with the Collector.
```

File: draft-pda-protocol.md [hunk 7]
```diff
@@ -146,13 +146,164 @@ document.
...
+{:br}
+
+
+In order to take a measurement, the collector first decides on the
+measurement parameters, including:
+
+* The values to be measured
+* The statistic to be computed (e.g., sum, mean, etc.)
+* The PPM scheme to use. This is to some extent dictated by the previous
+  choices.
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
* The values to be measured;
* The statistic to be computed (e.g., sum, mean, etc.); 
* The set of aggregators and necessary cryptographic keying material to use; and
* The PPM scheme to use. This is to some extent dictated by the previous
  choices.
```

File: draft-pda-protocol.md [hunk 8]
```diff
@@ -146,13 +146,164 @@ document.
...
+
+These parameters are distributed out of band to the clients and to
+the aggregators.
+
+During the duration of the measurement, each client records its own
+value(s), packages them up into a report and sends them to the leader.
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
value(s), packages them up into a report, and sends them to the leader.
```

File: draft-pda-protocol.md [hunk 9]
```diff
@@ -146,13 +146,164 @@ document.
...
+sharing. Rather than send its input in the clear, each client splits
+its measurements into a sequence of *shares* and sends a share to each
+of the aggregators. This provides two important properties:
+
+* It's impossible to deduce the measurement without knowing *all* of the shares.
+
+* It allows the aggregators to compute the final output by first
+   aggregating up their measurements shares locally, then combining
+   the results to obtain the final output.
+
```
Comment (chris-wood) [inline draft-pda-protocol.md:181]:
```suggestion
Aggregators also have the option of validating shares, as discussed in {{validating-inputs}}.
```

File: draft-pda-protocol.md [hunk 10]
```diff
@@ -146,13 +146,164 @@ document.
...
+The leader distributes the shares to the helpers and orchestrates
+the process of verifying them (see {{validating-inputs}})
+and assembling them into a final measurement for the collector.
+Depending on the PPM scheme, it may be possible to incrementally
+process each report as it comes in, or may be necesary to wait
+until all the required shares are received. This protocol is
+compatible with both a "push" mode in which the leader computes
+the measurements at predetermined points and sends the results
+to the collector and a "pull" mode in which the collector tells
+the leader when to perform the aggregation and return the results.
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
until all the required shares are received. 

This protocol is compatible with both a "push" deployment mode in which the 
leader computes the measurements at predetermined points and sends the results
to the collector and a "pull" mode in which the collector tells
the leader when to perform the aggregation and return the results.
```

Comment (chris-wood) [reply draft-pda-protocol.md]: Tried to clarify that modes are a deployment choice, and not something baked into the protocol.

File: draft-pda-protocol.md (hunk 9 above)
Comment (ekr) [reply draft-pda-protocol.md:181]: The option? Isn't it required effectively?

File: draft-pda-protocol.md (hunk 6 above)
Comment (ekr) [reply draft-pda-protocol.md]: I see what you want, but I think this is a bit confusing as written
"maybe as requested by the Collector"

Comment (chris-wood) [reply draft-pda-protocol.md]: Yeah, that's better

File: draft-pda-protocol.md (hunk 9 above)
Comment (chris-wood) [reply draft-pda-protocol.md:181]: There were some deployments that wanted to use this without actually verifying any proofs 🤷 , relying on client attestation and authentication of some form to ensure the data is trusted.

File: draft-pda-protocol.md [hunk 11]
```diff
@@ -592,113 +564,63 @@ encoded value of the following form:
...
-enough reports and the leader must have obtained the helper's encrypted output
-share (see {{pa-aggregate}}). In general, the procedure by which the aggregators
-verify and aggregate reports depends on parameters carried by the collect
-request.
+Once a set of clients have uploaded their reports to the leader, the leader
+can send them to the aggregators to be verified and aggregated. In order
+to enable the system to handle very large batches of reports, this process
+can be performed incrementally. To aggregate a set of reports,
+the leader sends an AggregateReq to the helper containing those report
+shares. The helper then processes the (verifying
```
Comment (chris-wood) [inline draft-pda-protocol.md]:
```suggestion
shares. The helper then processes them (verifying
```

**Pull request closed without merging**